### DISCUSSION TITLE
important notice for instance hosters and public v7 api users (api.cobalt.tools)

### DISCUSSION THREAD

**Discussion by wukko** [imputnet/cobalt#860] [announcements]
## what happened?
several bad actors are continuously scraping public cobalt instances for youtube videos. as a result, many (sometimes personal) instances either run out of bandwidth, cookies, or experience reduced performance. last week it got especially bad, completely knocking out several community instances.

in an attempt to stop this, i asked @hyperdefined and @KwiatekMiki to take down their instance lists. it didn't help much, as they've both been cached or saved via other means. the scraping continues, more instances become affected, hosters don't know what to do.

## what are we doing to stop this?
we've implemented several types of authentication that hosters can use to protect their instances. they can use turnstile, api keys, or both at the same time.

as for protecting the main instance, we've enabled turnstile and api keys. we plan on keeping it this way for the foreseeable future.

if you use the old v7 api (/api/json), **it will be shut down on november 11th, 2024**. unfortunately, scraping gets worse as cobalt becomes more known, and we can't keep up the free public api without impacting the regular users of [cobalt.tools](https://cobalt.tools). you can [host your own cobalt instance](https://github.com/imputnet/cobalt/blob/main/docs/run-an-instance.md) to keep using the cobalt api after that date. you can see a timeline for public api brownout in [this post](https://github.com/imputnet/cobalt/discussions/876).

instance lists will not go back up as-is. all future lists will be opt-in and instances must be added only after hoster's explicit consent to participate.

it really sucks that it has come down to this, but we have no other choice :(

## i have an instance, what can i do to protect it?
i made [an instruction document](https://github.com/imputnet/cobalt/blob/main/docs/protect-an-instance.md) that will guide you through the process of configuring cloudflare turnstile on your instance. it also covers the configuration of api keys in case you don't use your instance from a web interface.

if you're not using the latest cobalt api package and want to use all the security features, now is the perfect time to upgrade. aside from new features, you will also be able to use cobalt without hosting your own frontend instance. all you need to do is [add your processing instance domain in cobalt settings](https://cobalt.tools/settings/instances#community). support for api keys in web interface is coming very soon!

## who is scraping instances?
we have several suspects, but prefer to not reveal this information for sake of not giving them any publicity (that they clearly crave).

## better docs
as a part of actions taken, we updated all documentation for cobalt on github (including the main readme), it's easier than ever to read and navigate: https://github.com/imputnet/cobalt

## what does this all mean?
if you don't host an instance or don't use the cobalt api, you don't have to do anything and nothing will change for you!

[cobalt.tools](https://cobalt.tools) isn't affected and will keep working for the foreseeable future, nothing about its availability will change.

**Comment by wukko** [comment]
if you have any suggestions, feel free to leave them in comments here :3

**Reply by AngeloD2022** [reply]
You could maybe add a proof of work challenge to the API.

**Reply by wukko** [reply]
@AngeloD2022 are there any proof of work challenges that can’t be automatically solved by a js script outside of the page/browser?

**Reply by AngeloD2022** [reply]
Proof of work challenges are inherently automatically solved, but the point is to make the scraping operation as costly as possible for the threat actor, requiring a valid solution to the challenge for the service to work. Assuming they're doing it in mass, you could have some ephemeral memory store that relates IP (hashed in some way to protect privacy) -> number of API requests within the last 60 minutes. Based on that, you could increase the difficulty of this challenge. Eventually, it would get so difficult, that it becomes computationally infeasible to spam correctly.

**Reply by wukko** [reply]
scrapers usually use either ipv6 blocks or rotating residential proxies, ip can’t be the definitive factor here

**Reply by TechnoJo4** [reply]
You don't need to use IPs, just increase the work required when the server is under load.
Some websites do queue systems, forcing users to wait when there's too much load. This would be the same thing but without actually having to implement something as complicated as a queue system (PoW challenges can typically be implemented in a dozen or two lines of code with no memory cost on the server).
A user hopefully won't notice a few seconds of wait, might not even mind a minute of wait if the server's under load. But an attacker has to either lower their request rate according to how many CPU cores they have (and thus reduce load on the server) or throw money at more cores.

Well, if turnstile works, you hopefully won't need to bother, but proofs of work are an easy alternative to look into since they're so simple to implement.

**Reply by DataM0del** [reply]
turnstile is easy to automate. Everyone and their dog can easily solve or automate it, see https://github.com/Body-Alhoha/turnaround.

**Reply by wukko** [reply]
@DataM0del really? that repo is archived and last issues were about the script no longer working: https://github.com/Body-Alhoha/turnaround/issues/6

**Reply by DataM0del** [reply]
> @DataM0del really? that repo is archived and last issues were about the script no longer working: [Body-Alhoha/turnaround#6](https://github.com/Body-Alhoha/turnaround/issues/6)

Oh... maybe I'll make an updated instead if I get around to it and am not lazy.
For now, I guess you'll be fine with it.
It should be easy though because you just have to act like a human, so make smooth movements and random small turns to simulate actual mouse movements

**Reply by DataM0del** [reply]
Can't seem to get it run, probably a environment issue. Some debug logging is added by me because it wasn't working.
![image](https://github.com/user-attachments/assets/32dcb26f-e8ca-4145-a4ab-64058edfac6d)

**Reply by DataM0del** [reply]
Actually, it's probably because I don't have Firefox installed (I use Firefox Nightly)

**Reply by DataM0del** [reply]
![image](https://github.com/user-attachments/assets/89cd8021-2f55-4a70-88b0-12deaea711cf)
:|

**Reply by DataM0del** [reply]
Oh, cool.
![image](https://github.com/user-attachments/assets/e8b0631a-ed20-44e8-8eb6-357c69bdf209)

**Reply by DataM0del** [reply]
![image](https://github.com/user-attachments/assets/002a763f-df47-4c33-b21f-21389f166747)
I wonder how they could ever patch it... :skull:

**Reply by AngeloD2022** [reply]
> ![image](https://private-user-images.githubusercontent.com/183248792/379853345-002a763f-df47-4c33-b21f-21389f166747.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk3OTMxNjYsIm5iZiI6MTcyOTc5Mjg2NiwicGF0aCI6Ii8xODMyNDg3OTIvMzc5ODUzMzQ1LTAwMmE3NjNmLWRmNDctNGMzMy1iMjFmLTIxMzg5ZjE2Njc0Ny5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyNFQxODAxMDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02ZjA1MDJjODM0YjViOTM4MzVmNjg1ZGY3M2YyZDYxYmQxYzc2Mjg2MjQ1ZTk2MTI4YThjZjY0MTQxMDc5YmVhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.4z7b-Q3LtktVD7b43oZs6U0LyKXizoR6x5Fu2UqZdfs) I wonder how they could ever patch it... 💀

```js
Object.defineProperty(navigator.__proto__, 'webdriver', { get() { return false }})
```

**Reply by DataM0del** [reply]
Anyways, I've been working on it for a while, but I'm currently rewriting it to use undetected_chromedriver. That should fix the problem, or at least most of it.

**Reply by WhenTheBegininging** [reply]
fix cobalt

**Comment by wukko** [comment]
just posted the timeline for old api brownout: https://github.com/imputnet/cobalt/discussions/876

**Comment by Szprinktrap** [comment]
can't have nice things on the internet :(
at least cobalt is open source and selfhostable tho

**Comment by trainingdummy** [comment]
It seems I picked exactly the wrong day to discover this API. :joy:

**Comment by axellse** [comment]
even if you do host your own instance its not backwards compatible with the old api, is this something that could be added?

**Reply by wukko** [reply]
no, there won't be backwards compatibility with v7 schema, just update your app to follow the new (improved) schema: https://github.com/imputnet/cobalt/blob/main/docs/api.md

**Comment by hassaine688-glitch** [comment]
whaaat🥶🫥

**Comment by slimshad0** [comment]
thanks for project 💖💖💖💖💖
